File: contents/03/basic-1/solution.cpp
#include "bits/stdc++.h"
using namespace std;

struct Hash {
	vector<int>table;
	vector<bool>used;
	int mod;

	Hash(int _mod) {
		mod = _mod;
		table.assign(_mod, 0);
		used.assign(_mod, false);
	}
	void insert(int x) {
		for (int i = 0; i < mod; i++) {
			int idx = (x + i) % mod;
			if (!used[idx]) {
				used[idx] = true;
				table[idx] = x;
				return;
			}
			else if (table[idx] == x)return;
		}
	}
	bool search(int x) {
		for (int i = 0; i < mod; i++) {
			int idx = (x + i) % mod;
			if (!used[idx])return false;
			if (table[idx] == x)return true;
		}
		return false;
	}
};

void solve() {
	Hash hash(100003);
	int q; cin >> q;
	while (q--) {
		int sign, x; cin >> sign >> x;
		if (sign == 0) {
			hash.insert(x);
		}
		else {
			if (hash.search(x))cout << "found" << endl;
			else cout << "not found" << endl;
		}
	}
}
signed main() {
	solve();
}

File: contents/03/basic-1/solution.py
class Hash:
    def __init__(self, _mod):
        self.mod = _mod
        self.used = [False]*_mod
        self.table = [0]*_mod
    
    def insert(self, x):
        #開番地を線形走査する
        for i in range(self.mod):
            idx = (x + i) % self.mod
            if not self.used[idx]:
                self.used[idx] = True
                self.table[idx] = x
                return
            elif self.table[idx] == x:
                return
    
    def search(self, x):
        for i in range(self.mod):
            idx = (x + i) % self.mod
            if not self.used[idx]:
                return False
            if self.table[idx] == x:
                return True
        return False


Q = int(input())
hash = Hash(100003)
for _ in range(Q):
    sign, x = map(int,input().split())
    if sign == 0:
        hash.insert(x)
    else:
        if hash.search(x):
            print("found")
        else:
            print("not found")
